README.md
NetApp E-Series SANtricity Collection
=========
    The SANtricity collection consist of the latest available versions of the NetApp E-Series SANtricity modules and roles.

        This collection provides NetApp E-Series customers with a wide range of configuration options through the collection's modules. However, the real
    benefit of using the SANtricity collection is found in the host and management roles. These roles provide ready-made, policy-based orchestration for
    E-Series platforms based on predefined role variables.
        Once the physical hardware has been installed, the SANtricity roles are capable of discovering the DHCP-assigned addresses, setting initial passwords
    and management interfaces so your automation can do full deployments for you without logging directly into the devices. Yet that's just the beginning,
    the management role will also ensure alerts, ASUP, logging, LDAP, and firmware are configured as expected; and the host role will setup host interfaces,
    provision and map storage, and if your servers are defined in the inventory, correctly populate E-Series hosts and host groups automatically.

    Roles:
        - nar_santricity_common: Discover NetApp E-Series storage systems and configures SANtricity Web Services Proxy.
        - nar_santricity_host: Configure storage pools, volumes, hosts, host groups, and port interfaces.
        - nar_santricity_management: Manage storage system's name, management interfaces, alerts, syslog, auditlog, asup, ldap, certificates, drive firmware and controller firmware.
            
    Modules:
        - na_santricity_alerts: Manage email alert notification settings
        - na_santricity_alerts_syslog: Manage syslog servers receiving storage system alerts
        - na_santricity_asup: Manage auto-support settings
        - na_santricity_auditlog: Manage audit-log configuration
        - na_santricity_auth: Set or update the password for a storage array
        - na_santricity_client_certificate: Manage remote server certificates
        - na_santricity_discover: Discover E-Series storage systems on a subnet
        - na_santricity_drive_firmware: Manage drive firmware
        - na_santricity_facts: Retrieve facts about NetApp E-Series storage arrays
        - na_santricity_firmware: Manage firmware
        - na_santricity_global: Manage global settings configuration
        - na_santricity_host: Manage eseries hosts
        - na_santricity_hostgroup: Manage array host groups
        - na_santricity_iscsi_interface: Manage iSCSI interface configuration
        - na_santricity_iscsi_target: Manage iSCSI target configuration
        - na_santricity_ldap: Manage LDAP integration to use for authentication
        - na_santricity_lun_mapping: Manage lun mappings
        - na_santricity_mgmt_interface: Manage management interface configuration
        - na_santricity_storage_system: Manage SANtricity web services proxy storage arrays
        - na_santricity_storagepool: Manage volume groups and disk pools
        - na_santricity_syslog: Manage syslog settings
        - na_santricity_volume: Manage storage volumes
        
    Deprecated Modules:
        - netapp_e_alerts: Manage email notification settings
        - netapp_e_amg: Create, remove, and update asynchronous mirror groups
        - netapp_e_amg_role: Update the role of a storage array within an Asynchronous Mirror Group (AMG)
        - netapp_e_amg_sync: Conduct synchronization actions on asynchronous mirror groups
        - netapp_e_asup: Manage auto-support settings
        - netapp_e_auditlog: Manage audit-log configuration
        - netapp_e_auth: Set or update the password for a storage array
        - netapp_e_drive_firmware: Manage drive firmware
        - netapp_e_facts: Retrieve facts about NetApp E-Series storage arrays
        - netapp_e_firmware: Manage firmware
        - netapp_e_flashcache: Manage SSD caches
        - netapp_e_global: Manage global settings configuration
        - netapp_e_hostgroup: Manage eseries hosts
        - netapp_e_host: Manage array host groups
        - netapp_e_iscsi_interface: Manage iSCSI interface configuration
        - netapp_e_iscsi_target: Manage iSCSI target configuration
        - netapp_e_ldap: Manage LDAP integration to use for authentication
        - netapp_e_lun_mapping: Create, delete, or modify lun mappings
        - netapp_e_mgmt_interface: Manage management interface configuration
        - netapp_e_snapshot_group: Manage snapshot groups
        - netapp_e_snapshot_images: Create and delete snapshot images
        - netapp_e_snapshot_volume: Manage snapshot volumes
        - netapp_e_storagepool: Manage volume groups and disk pools
        - netapp_e_storage_system: Manage Web Services Proxy manage storage arrays
        - netapp_e_syslog: Manage syslog settings
        - netapp_e_volume_copy: Create volume copy pairs
        - netapp_e_volume: Manage storage volumes (standard and thin)

Requirements
------------
    - Ansible 2.9 or later
    - NetApp E-Series E2800 platform or newer or NetApp E-Series SANtricity Web Services Proxy configured for older E-Series Storage arrays.

Example Playbook
----------------
    - hosts: eseries_storage_systems
      gather_facts: false
      collections:
        - netapp_eseries.santricity
      tasks:
        - name: Ensure proxy has been configured and storage systems have been discovered.
          import_role:
            name: nar_santricity_common
        - name: Ensure all management related policies are enforced.
          import_role:
            name: nar_santricity_management
        - name: Ensure all host related policies are enforced.
          import_role:
            name: nar_santricity_host

Example Storage System Inventory File (Simple example without storage system discovery)
-------------------------------------
    eseries_system_api_url: https://192.168.1.200:8443/devmgr/v2/
    eseries_system_password: admin_password
    eseries_validate_certs: false

    eseries_system_name: my_eseries_array

    eseries_management_interfaces:
      config_method: static
      subnet_mask: 255.255.255.0
      gateway: 192.168.1.1
      dns_config_method: static
      dns_address: 192.168.1.253
      ntp_config_method: static
      ntp_address: 192.168.1.200
      controller_a:
        - address: 192.168.1.100
        - address: 192.168.1.101
      controller_b:
        - address: 192.168.1.102
        - address: 192.168.1.103

    eseries_initiator_protocol: fc

    # Storage pool and volume configuration
    eseries_storage_pool_configuration:
      - name: vg_01
        raid_level: raid6
        criteria_drive_count: 10
        volumes:
          - name: metadata
            host: servers
            size: 4096
      - name: vg_02
        raid_level: raid5
        criteria_drive_count: 10
        volumes:
          - name: storage_[1-3]
            host: servers
            size: 2
            size_unit: tb
            volume_metadata:  # Used by netapp_eseries.host.mount role to format and mount volumes
              format_type: xfs
              mount_options1: "noatime,nodiratime,logbufs=8,logbsize=256k,largeio"
              mount_options2: "inode64,swalloc,allocsize=131072k,nobarrier,_netdev"
              mount_directory: "/data/beegfs/"

Example Storage System Inventory File (Discover storage system)
-------------------------------------
**Note that this discovery method works for SANtricity versions 11.62 or later, otherwise it will only discover the systems with unset passwords.**

    eseries_system_serial: "012345678901"   # Be sure to quote if the serial is all numbers and begins with zero.
    eseries_system_password: admin_password
    eseries_subnet: 192.168.1.0/24

    eseries_management_interfaces:          # (Optional) Specifying static management interfaces can be used not only to discover the storage system but also to contact when valid.
      config_method: static
      controller_a:
        - address: 192.168.1.100
      controller_b:
        - address: 192.168.1.101

    (...)   # Same as the previous examples starting with the eseries_validate_certs line

Example Storage System Inventory File (Discover storage system with proxy)
-------------------------------------
    eseries_system_serial: "012345678901"   # Be sure to quote if the serial is all numbers and begins with zero.
    eseries_system_password: admin_password

    eseries_proxy_api_url: https://192.168.1.100:8443/devmgr/v2/
    eseries_proxy_api_password: admin_password
    eseries_subnet: 192.168.1.0/24
    eseries_prefer_embedded: false    # Overrides the default behavior of using Web Services Proxy when eseries_proxy_api_url is defined. This will only effect
                                      #     storage systems that have Embedded Web Services.

    (...)   # Same as the previous examples starting with the eseries_validate_certs line

Example Storage System Inventory File
-------------------------------------
    eseries_system_api_url: https://192.168.1.200:8443/devmgr/v2/
    eseries_system_password: admin_password
    eseries_validate_certs: false

    eseries_system_name: my_eseries_array
    eseries_system_cache_block_size: 128
    eseries_system_cache_flush_threshold: 90
    eseries_system_autoload_balance: enabled
    eseries_system_host_connectivity_reporting: enabled
    eseries_system_default_host_type: Linux DM-MP

    eseries_management_interfaces:
      config_method: static
      subnet_mask: 255.255.255.0
      gateway: 192.168.1.1
      dns_config_method: static
      dns_address: 192.168.1.253
      dns_address_backup:  192.168.1.254
      ssh: true
      ntp_config_method: static
      ntp_address: 192.168.1.200
      ntp_address_backup: 192.168.1.201
      controller_a:
        - address: 192.168.1.100
        - address: 192.168.1.101
      controller_b:
        - address: 192.168.1.102
        - address: 192.168.1.103

    eseries_ldap_state: present
    eseries_ldap_bind_username:
    eseries_ldap_bind_password:
    eseries_ldap_server:
    eseries_ldap_search_base:
    eseries_ldap_role_mappings:
      ".*":
        - storage.admin
        - storage.monitor
        - support.admin
        - security.admin

    eseries_client_certificate_certificates:
      - /path/to/client_certificate.crt

    eseries_firmware_firmware: "/path/to/firmware.dlp"
    eseries_firmware_nvsram: "/path/to/nvsram.dlp"
    eseries_drive_firmware_firmware_list:
      - "/path/to/drive_firmware.dlp"

    eseries_asup_state: enabled
    eseries_asup_active: true
    eseries_asup_days: [sunday, saturday]
    eseries_asup_start: 17
    eseries_asup_end: 24
    eseries_asup_validate: false
    eseries_asup_method: email
    eseries_asup_email:
      server: smtp.example.com
      sender: noreply@example.com

    eseries_syslog_state: present
    eseries_syslog_address: 192.168.1.150
    eseries_syslog_protocol: udp
    eseries_syslog_port: 514
    eseries_alert_syslog_servers:
        - "address": 192.168.1.150
          "port": 514
    eseries_initiator_protocol: iscsi
    
    # Controller port definitions
    eseries_controller_iscsi_port_config_method: static
    eseries_controller_iscsi_port_subnet_mask: 255.255.255.0
    eseries_controller_iscsi_port:
      controller_a:
        - address: 192.168.2.100
        - address: 192.168.2.110
      controller_b:
        - address: 192.168.3.100
        - address: 192.168.3.110

    # Storage pool and volume configuration
    eseries_storage_pool_configuration:
      - name: pool[1-2]
        raid_level: raid6
        criteria_drive_count: 10
        volumes:
          - name: "[pool]_volume[A-C]"
            host: server_group
            size: 10
            size_unit: tb

Collection Variables
--------------
**Note that when values are specified below, they indicate the default value.**

    eseries_subnet:                   # Network subnet to search for the storage system specified in CIDR form. Example: 192.168.1.0/24
    eseries_template_api_url:         # Template for the web services api url. Default: https://0.0.0.0:8443/devmgr/v2/
    eseries_prefer_embedded: false    # Overrides the default behavior of using Web Services Proxy when eseries_proxy_api_url is defined. This will only effect
                                      #     storage systems that have Embedded Web Services.
    eseries_validate_certs: true      # Indicates Whether SSL certificates should be verified. Used for both embedded and proxy. Choices: true, false

    # Storage system specific variables
    eseries_proxy_ssid:               # Arbitrary string for the proxy to represent the storage system. eseries_system_serial will be used when not defined.
    eseries_system_serial:            # Storage system serial number. Be sure to quote if the serial is all numbers and begins with zero. (This is located on a
                                      #     label at the top-left towards the front on the device)
    eseries_system_addresses:         # Storage system management IP addresses. Only required when eseries_system_serial or eseries_system_api_url are not
                                      #     defined. When not specified, addresses will be populated with eseries_management_interfaces controller addresses.
    eseries_system_api_url:           # Url for the storage system's for embedded web services rest api. Example: https://192.168.10.100/devmgr/v2
    eseries_system_username: admin    # Username for the storage system's for embedded web services rest api
    eseries_system_password:          # Password for the storage system's for embedded web services rest api and when the admin password has not been set
                                      #     eseries_system_password will be used to set it.
    eseries_system_tags:              # Meta tags to associate with storage system when added to the proxy.

    # Web Services Proxy specific variable
        Note: eseries_proxy_* variables are required to discover storage systems prior to SANtricity OS version 11.60.2.
    eseries_proxy_api_url:                  # Url for the storage system's for proxy web services rest api. Example: https://192.168.10.100/devmgr/v2
    eseries_proxy_api_username: admin       # Username for the storage system's for proxy web services rest api.
    eseries_proxy_api_password:             # Password for the storage system's for proxy web services rest api and when the admin password has not been set
                                            #   eseries_proxy_api_password will be used to set it.
    eseries_proxy_monitor_password:         # Proxy password for the monitor username
    eseries_proxy_security_password:        # Proxy password for the security username
    eseries_proxy_storage_password:         # Proxy password for the monitor username
    eseries_proxy_support_password:         # Proxy password for the support username
    eseries_proxy_accept_certifications:    # Force automatic acceptance of all storage system's certificate
    eseries_proxy_default_system_tags:      # Default meta tags to associate with all storage systems
    eseries_proxy_default_password:         # Default password to associate with all storage systems. This is overridden by eseries_system_password.

    # LDAP configuration defaults
    eseries_proxy_ldap_state:             # Whether LDAP should be configured for the proxy`
    eseries_proxy_ldap_identifier:        # The user attributes that should be considered for the group to role mapping
    eseries_proxy_ldap_user_attribute:    # Attribute used to the provided username during authentication.
    eseries_proxy_ldap_bind_username:     # User account that will be used for querying the LDAP server.
    eseries_proxy_ldap_bind_password:     # Password for the bind user account
    eseries_proxy_ldap_server:            # LDAP server URL.
    eseries_proxy_ldap_search_base:       # Search base used for find user's group membership
    eseries_proxy_ldap_role_mappings:     # Dictionary of user groups, each containing the list of access roles.
                                          #     Role choices: storage.admin - allows users full read/writes access to storage objects and operations.
                                          #                   storage.monitor - allows users read-only access to storage objects and operations.
                                          #                   storage.admin - allows users access to hardware, diagnostic information, major event logs,
                                          #                       and other critical support-related functionality, but not the sorage configuration.
                                          #                   security.admin - allows users access to authentication/authorization configuration, as
                                          #                       well as the audit log configuration, adn certification management.

    # Global storage system information
    eseries_system_name:                           # Name of the storage system.
    eseries_system_cache_block_size:               # Cache block size
    eseries_system_cache_flush_threshold:          # Unwritten data will be flushed when exceeds this threshold
    eseries_system_autoload_balance:               # Whether automatic load balancing should be enabled. Choices: enabled, disabled
    eseries_system_host_connectivity_reporting:    # Whether host connectivity reporting should be enabled. Choices: enabled, disabled
    eseries_system_default_host_type:              # Only required when using something other than Linux kernel 3.10 or later with DM-MP (Linux DM-MP),
                                                   #    non-clustered Windows (Windows), or the storage system default host type is incorrect.
                                                   # Common host type definitions:
                                                   #     - AIX MPIO: The Advanced Interactive Executive (AIX) OS and the native MPIO driver
                                                   #     - AVT 4M: Silicon Graphics, Inc. (SGI) proprietary multipath driver
                                                   #     - HP-UX: The HP-UX OS with native multipath driver
                                                   #     - Linux ATTO: The Linux OS and the ATTO Technology, Inc. driver (must use ATTO FC HBAs)
                                                   #     - Linux DM-MP: The Linux OS and the native DM-MP driver
                                                   #     - Linux Pathmanager: The Linux OS and the SGI proprietary multipath drive
                                                   #     - Mac: The Mac OS and the ATTO Technology, Inc. driver
                                                   #     - ONTAP: FlexArray
                                                   #     - Solaris 11 or later: The Solaris 11 or later OS and the native MPxIO driver
                                                   #     - Solaris 10 or earlier: The Solaris 10 or earlier OS and the native MPxIO driver
                                                   #     - SVC: IBM SAN Volume Controller
                                                   #     - VMware: ESXi OS
                                                   #     - Windows: Windows Server OS and Windows MPIO with a DSM driver
                                                   #     - Windows Clustered: Clustered Windows Server OS and Windows MPIO with a DSM driver
                                                   #     - Windows ATTO: Windows OS and the ATTO Technology, Inc. driver

    # Role-based username passwords
    eseries_system_monitor_password:     # Storage system monitor username password
    eseries_system_security_password:    # Storage system security username password
    eseries_system_storage_password:     # Storage system storage username password
    eseries_system_support_password:     # Storage system support username password

    # Storage management interface defaults
        Note:  eseries_management_* variables have the lowest priority and will be overwritten by those found in eseries_management_interfaces
    eseries_management_config_method:         # Default config method for all management interfaces. Choices: static, dhcp
    eseries_management_subnet_mask:           # Default subnet mask for all management interfaces
    eseries_management_gateway_mask:          # Default gateway for all management interfaces
    eseries_management_dns_config_method:     # Default DNS config method for all management interfaces
    eseries_management_dns_address:           # Default primary DNS address for all management interfaces

CHANGELOG.md
## [1.0.1] - 2020-03-016
### Added
This release consists of all existing NetApp E-Series Ansible content modified to conform with Ansible collections.

Initial Collection Content:
 - Roles:
    - nar_santricity_common: Discover NetApp E-Series storage systems and configures SANtricity Web Services Proxy.
    - nar_santricity_host: Configure storage pools, volumes, hosts, host groups, and port interfaces. (nar_santricity_common is called at the beginning)
    - nar_santricity_management: Manage storage system's name, management interfaces, alerts, syslog, auditlog, asup, ldap, certificates, drive firmware and
        controller firmware. (nar_santricity_common is called at the beginning)
 - Modules:
    - na_santricity_alerts: Manage email alert notification settings
    - na_santricity_alerts_syslog: Manage syslog servers receiving storage system alerts
    - na_santricity_asup: Manage auto-support settings
    - na_santricity_auditlog: Manage audit-log configuration
    - na_santricity_auth: Set or update the password for a storage array
    - na_santricity_client_certificate: Manage remote server certificates
    - na_santricity_discover: Discover E-Series storage systems on a subnet
    - na_santricity_drive_firmware: Manage drive firmware
    - na_santricity_facts: Retrieve facts about NetApp E-Series storage arrays
    - na_santricity_firmware: Manage firmware
    - na_santricity_global: Manage global settings configuration
    - na_santricity_host: Manage eseries hosts
    - na_santricity_hostgroup: Manage array host groups
    - na_santricity_iscsi_interface: Manage iSCSI interface configuration
    - na_santricity_iscsi_target: Manage iSCSI target configuration
    - na_santricity_ldap: Manage LDAP integration to use for authentication
    - na_santricity_lun_mapping: Manage lun mappings
    - na_santricity_mgmt_interface: Manage management interface configuration
    - na_santricity_storage_system: Manage SANtricity web services proxy storage arrays
    - na_santricity_storagepool: Manage volume groups and disk pools
    - na_santricity_syslog: Manage syslog settings
    - na_santricity_volume: Manage storage volumes
 - Deprecated Modules:
    - netapp_e_alerts: Manage email notification settings
    - netapp_e_amg: Create, remove, and update asynchronous mirror groups
    - netapp_e_amg_role: Update the role of a storage array within an Asynchronous Mirror Group (AMG)
    - netapp_e_amg_sync: Conduct synchronization actions on asynchronous mirror groups
    - netapp_e_asup: Manage auto-support settings
    - netapp_e_auditlog: Manage audit-log configuration
    - netapp_e_auth: Set or update the password for a storage array
    - netapp_e_drive_firmware: Manage drive firmware
    - netapp_e_facts: Retrieve facts about NetApp E-Series storage arrays
    - netapp_e_firmware: Manage firmware
    - netapp_e_flashcache: Manage SSD caches
    - netapp_e_global: Manage global settings configuration
    - netapp_e_hostgroup: Manage eseries hosts
    - netapp_e_host: Manage array host groups
    - netapp_e_iscsi_interface: Manage iSCSI interface configuration
    - netapp_e_iscsi_target: Manage iSCSI target configuration
    - netapp_e_ldap: Manage LDAP integration to use for authentication
    - netapp_e_lun_mapping: Create, delete, or modify lun mappings
    - netapp_e_mgmt_interface: Manage management interface configuration
    - netapp_e_snapshot_group: Manage snapshot groups
    - netapp_e_snapshot_images: Create and delete snapshot images
    - netapp_e_snapshot_volume: Manage snapshot volumes
    - netapp_e_storagepool: Manage volume groups and disk pools
    - netapp_e_storage_system: Manage Web Services Proxy manage storage arrays
    - netapp_e_syslog: Manage syslog settings
    - netapp_e_volume_copy: Create volume copy pairs


plugins/modules/netapp_e_asup.py
#!/usr/bin/python

# (c) 2018, NetApp, Inc
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type

ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['deprecated'],
                    'supported_by': 'community'}

DOCUMENTATION = """
---
module: netapp_e_asup
short_description: NetApp E-Series manage auto-support settings
description:
    - Allow the auto-support settings to be configured for an individual E-Series storage-system
version_added: '2.7'
author: Michael Price (@lmprice)
extends_documentation_fragment:
    - netapp_eseries.santricity.santricity.netapp.eseries
options:
    state:
        description:
            - Enable/disable the E-Series auto-support configuration.
            - When this option is enabled, configuration, logs, and other support-related information will be relayed
              to NetApp to help better support your system. No personally identifiable information, passwords, etc, will
              be collected.
        default: enabled
        choices:
            - enabled
            - disabled
        aliases:
            - asup
            - auto_support
            - autosupport
    active:
        description:
            - Enable active/proactive monitoring for ASUP. When a problem is detected by our monitoring systems, it's
              possible that the bundle did not contain all of the required information at the time of the event.
              Enabling this option allows NetApp support personnel to manually request transmission or re-transmission
              of support data in order ot resolve the problem.
            - Only applicable if I(state=enabled).
        default: yes
        type: bool
    start:
        description:
            - A start hour may be specified in a range from 0 to 23 hours.
            - ASUP bundles will be sent daily between the provided start and end time (UTC).
            - I(start) must be less than I(end).
        aliases:
            - start_time
        default: 0
    end:
        description:
            - An end hour may be specified in a range from 1 to 24 hours.
            - ASUP bundles will be sent daily between the provided start and end time (UTC).
            - I(start) must be less than I(end).
        aliases:
            - end_time
        default: 24
    days:
        description:
            - A list of days of the week that ASUP bundles will be sent. A larger, weekly bundle will be sent on one
              of the provided days.
        choices:
            - monday
            - tuesday
            - wednesday
            - thursday
            - friday
            - saturday
            - sunday
        required: no
        aliases:
            - days_of_week
            - schedule_days
    verbose:
        description:
            - Provide the full ASUP configuration in the return.
        default: no
        required: no
        type: bool
    log_path:
        description:
            - A local path to a file to be used for debug logging
        required: no
notes:
    - Check mode is supported.
    - Enabling ASUP will allow our support teams to monitor the logs of the storage-system in order to proactively
      respond to issues with the system. It is recommended that all ASUP-related options be enabled, but they may be
      disabled if desired.
    - This API is currently only supported with the Embedded Web Services API v2.0 and higher.
"""

EXAMPLES = """
    - name: Enable ASUP and allow pro-active retrieval of bundles
      netapp_e_asup:
        state: enabled
        active: yes
        api_url: "10.1.1.1:8443"
        api_username: "admin"
        api_password: "myPass"

    - name: Set the ASUP schedule to only send bundles from 12 AM CST to 3 AM CST.
      netapp_e_asup:
        start: 17
        end: 20
        api_url: "10.1.1.1:8443"
        api_username: "admin"
        api_password: "myPass"
"""

RETURN = """
msg:
    description: Success message
    returned: on success
    type: str
    sample: The settings have been updated.
asup:
    description:
        - True if ASUP is enabled.
    returned: on success
    sample: True
    type: bool
active:
    description:
        - True if the active option has been enabled.
    returned: on success
    sample: True
    type: bool
cfg:
    description:
        - Provide the full ASUP configuration.
    returned: on success when I(verbose=true).
    type: complex
    contains:
        asupEnabled:
            description:
                    - True if ASUP has been enabled.
            type: bool
        onDemandEnabled:
            description:
                    - True if ASUP active monitoring has been enabled.
            type: bool
        daysOfWeek:
            description:
                - The days of the week that ASUP bundles will be sent.
            type: list
"""

import json
import logging
from pprint import pformat

from ansible.module_utils.basic import AnsibleModule
from ansible_collections.netapp_eseries.santricity.plugins.module_utils.netapp import request, eseries_host_argument_spec
from ansible.module_utils._text import to_native

HEADERS = {
    "Content-Type": "application/json",
    "Accept": "application/json",
}


class Asup(object):
    DAYS_OPTIONS = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday']

    def __init__(self):
        argument_spec = eseries_host_argument_spec()
        argument_spec.update(dict(
            state=dict(type='str', required=False, default='enabled', aliases=['asup', 'auto_support', 'autosupport'],
                       choices=['enabled', 'disabled']),
            active=dict(type='bool', required=False, default=True, ),
            days=dict(type='list', required=False, aliases=['schedule_days', 'days_of_week'],
                      choices=self.DAYS_OPTIONS),
            start=dict(type='int', required=False, default=0, aliases=['start_time']),
            end=dict(type='int', required=False, default=24, aliases=['end_time']),
            verbose=dict(type='bool', required=False, default=False),
            log_path=dict(type='str', required=False),
        ))

        self.module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True, )
        args = self.module.params
        self.asup = args['state'] == 'enabled'
        self.active = args['active']
        self.days = args['days']
        self.start = args['start']
        self.end = args['end']
        self.verbose = args['verbose']

        self.ssid = args['ssid']
        self.url = args['api_url']
        self.creds = dict(url_password=args['api_password'],
                          validate_certs=args['validate_certs'],
                          url_username=args['api_username'], )

        self.check_mode = self.module.check_mode

        log_path = args['log_path']

        # logging setup
        self._logger = logging.getLogger(self.__class__.__name__)

        if log_path:
            logging.basicConfig(
                level=logging.DEBUG, filename=log_path, filemode='w',
                format='%(relativeCreated)dms %(levelname)s %(module)s.%(funcName)s:%(lineno)d\n %(message)s')

        if not self.url.endswith('/'):
            self.url += '/'

        if self.start >= self.end:
            self.module.fail_json(msg="The value provided for the start time is invalid."
                                      " It must be less than the end time.")
        if self.start < 0 or self.start > 23:
            self.module.fail_json(msg="The value provided for the start time is invalid. It must be between 0 and 23.")
        else:
            self.start = self.start * 60
        if self.end < 1 or self.end > 24:
            self.module.fail_json(msg="The value provided for the end time is invalid. It must be between 1 and 24.")
        else:
            self.end = min(self.end * 60, 1439)

        if not self.days:
            self.days = self.DAYS_OPTIONS

    def get_configuration(self):
        try:
            (rc, result) = request(self.url + 'device-asup', headers=HEADERS, **self.creds)

            if not (result['asupCapable'] and result['onDemandCapable']):
                self.module.fail_json(msg="ASUP is not supported on this device. Array Id [%s]." % (self.ssid))
            return result

        except Exception as err:
            self.module.fail_json(msg="Failed to retrieve ASUP configuration! Array Id [%s]. Error [%s]."
                                      % (self.ssid, to_native(err)))

    def update_configuration(self):
        config = self.get_configuration()
        update = False
        body = dict()

        if self.asup:
            body = dict(asupEnabled=True)
            if not config['asupEnabled']:
                update = True

            if (config['onDemandEnabled'] and config['remoteDiagsEnabled']) != self.active:
                update = True
                body.update(dict(onDemandEnabled=self.active,
                                 remoteDiagsEnabled=self.active))
            self.days.sort()
            config['schedule']['daysOfWeek'].sort()

            body['schedule'] = dict(daysOfWeek=self.days,
                                    dailyMinTime=self.start,
                                    dailyMaxTime=self.end,
                                    weeklyMinTime=self.start,
                                    weeklyMaxTime=self.end)

            if self.days != config['schedule']['daysOfWeek']:
                update = True
            if self.start != config['schedule']['dailyMinTime'] or self.start != config['schedule']['weeklyMinTime']:
                update = True
            elif self.end != config['schedule']['dailyMaxTime'] or self.end != config['schedule']['weeklyMaxTime']:
                update = True

        elif config['asupEnabled']:
            body = dict(asupEnabled=False)
            update = True

        self._logger.info(pformat(body))

        if update and not self.check_mode:
            try:
                (rc, result) = request(self.url + 'device-asup', method='POST',
                                       data=json.dumps(body), headers=HEADERS, **self.creds)
            # This is going to catch cases like a connection failure
            except Exception as err:
                self.module.fail_json(msg="We failed to set the storage-system name! Array Id [%s]. Error [%s]."
                                          % (self.ssid, to_native(err)))

        return update

    def update(self):
        update = self.update_configuration()
        cfg = self.get_configuration()
        if self.verbose:
            self.module.exit_json(msg="The ASUP settings have been updated.", changed=update,
                                  asup=cfg['asupEnabled'], active=cfg['onDemandEnabled'], cfg=cfg)
        else:
            self.module.exit_json(msg="The ASUP settings have been updated.", changed=update,
                                  asup=cfg['asupEnabled'], active=cfg['onDemandEnabled'])

    def __call__(self, *args, **kwargs):
        self.update()


def main():
    settings = Asup()
    settings()


if __name__ == '__main__':
    main()
